decimal.go

```diff
@@ -773,12 +774,27 @@ func (d *Decimal) Scan(src interface{}) error {
 	}
 }
 
+var jsonNull = []byte("null")
+
+// UnmarshalJSON implements the json.Unmarshaler interface.
+func (d *Decimal) UnmarshalJSON(b []byte) error {
+	if bytes.Compare(b, jsonNull) == 0 {
```

Comment:
Let's leave this as `if string(text) == "null"` like it is for math/big. I think the compiler is smart enough to optimize it.

Reply:
AFAIK, this will result in a memory copy every time since strings are immutable but bytes slices are mutable. The incoming `[]bytes` aren't static (depends on the JSON being unmarshaled) so the compiler can't optimize this.

Reply:
Please change it to the string version. Here's a few reasons:

- The compiler is pretty smart and if I had to bet, I would say that it's smart enough to avoid the string -> bytes copy each call.
- The function call into bytes.Compare isn't cheap.
- The go stdlib already does this, and I have no reason to not copy them.
- My benchmarks show that the string version is faster:

```
func BenchmarkBytes(b *testing.B) {
	var bb = []byte("blah")
	var a bool
	for i := 0; i < b.N; i++ {
		a = bytes.Compare(bb, bb) == 0
	}
	if !a {
		b.Fatal("balh")
	}
}

func BenchmarkString(b *testing.B) {
	var xb = []byte("blah")
	var a bool
	for i := 0; i < b.N; i++ {
		a = string(xb) == "blah"
	}
	if !a {
		b.Fatal("balh")
	}
}
```

```

BenchmarkBytes-12     	500000000	         3.14 ns/op
BenchmarkString-12    	2000000000	         0.62 ns/op
```

My mind could be changed by any evidence, like a benchmark or assembly output analysis that show differently, but as it is there's no evidence that the bytes version is better.

Reply:
Odd, I get similar benchmark performances even with the following changes:
1. Use `bytes.Equal()` instead of `bytes.Compare`
1. Use a table of byte slices to force string copying conversion

```
package comparison

import (
	"bytes"
	"math/rand"
	"testing"
	"time"
)

var sampleBytes = [][]byte{
	[]byte("blah"),
	[]byte(""),
	[]byte("null"),
	[]byte("foo"),
	[]byte("bar"),
	[]byte("foobar"),
	[]byte("hello"),
	[]byte("world"),
	[]byte("helloworld"),
	[]byte("deadbeef"),
}

const letterBytes = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
const (
	letterIdxBits = 6                    // 6 bits to represent a letter index
	letterIdxMask = 1<<letterIdxBits - 1 // All 1-bits, as many as letterIdxBits
	letterIdxMax  = 63 / letterIdxBits   // # of letter indices fitting in 63 bits
)

var src = rand.NewSource(time.Now().UnixNano())

func RandBytes(n int) []byte {
	b := make([]byte, n)
	// A src.Int63() generates 63 random bits, enough for letterIdxMax characters!
	for i, cache, remain := n-1, src.Int63(), letterIdxMax; i >= 0; {
		if remain == 0 {
			cache, remain = src.Int63(), letterIdxMax
		}
		if idx := int(cache & letterIdxMask); idx < len(letterBytes) {
			b[i] = letterBytes[idx]
			i--
		}
		cache >>= letterIdxBits
		remain--
	}
	return b
}

func init() {
	for i := 0; i < 100; i++ {
		sampleBytes = append(sampleBytes, RandBytes(100))
	}
}

func BenchmarkBytes(b *testing.B) {
	var bb = []byte("blah")
	var a bool
	for i := 0; i < b.N; i++ {
		j := i % len(sampleBytes)
		a = bytes.Equal(bb, sampleBytes[j])
	}
	b.Log(a)
}

func BenchmarkString(b *testing.B) {
	var xb = []byte("blah")
	var a bool
	for i := 0; i < b.N; i++ {
		j := i % len(sampleBytes)
		a = string(xb) == string(sampleBytes[j])
	}
	b.Log(a)
}
```
```
goos: darwin
goarch: amd64
BenchmarkBytes-8    	100000000	        13.2 ns/op
--- BENCH: BenchmarkBytes-8
	comparison_test.go:62: true
	comparison_test.go:62: false
	comparison_test.go:62: false
	comparison_test.go:62: false
	comparison_test.go:62: false
BenchmarkString-8   	100000000	        10.4 ns/op
--- BENCH: BenchmarkString-8
	comparison_test.go:72: true
	comparison_test.go:72: false
	comparison_test.go:72: false
	comparison_test.go:72: false
	comparison_test.go:72: false
PASS
```

So I'll change to the string conversion.